### ISSUE TITLE
Pull request jobs are flaking with k3s

### ISSUE THREAD

**Issue by jsturtevant** [containerd/runwasi#322]
Wasmer has failed in e2es in several PR's recently:

https://github.com/containerd/runwasi/pull/320
https://github.com/containerd/runwasi/pull/319
https://github.com/containerd/runwasi/pull/318

Looking at logs it doesn't give a ton of info, looks like the pods are stuck in pending state

fyi @0xE282B0 @dierbei

**Comment by jprendes** [comment]
I'm not sure it's related to warmer, I think it's rather flakiness from k3s.
* [#1191](https://github.com/containerd/runwasi/actions/runs/6225493086): wasmer
* [#1190](https://github.com/containerd/runwasi/actions/runs/6225491702): wasmer
* [#1189](https://github.com/containerd/runwasi/actions/runs/6225488999): wasmer
* [#1186](https://github.com/containerd/runwasi/actions/runs/6203705560): wasmedge
* [#1185](https://github.com/containerd/runwasi/actions/runs/6203644167): wasmer
* [#1184](https://github.com/containerd/runwasi/actions/runs/6203583693): wasmedge
* [#1182](https://github.com/containerd/runwasi/actions/runs/6202982553): wasmedge
* [#1181](https://github.com/containerd/runwasi/actions/runs/6202883315): wasmer + wasmedge
* [#1175](https://github.com/containerd/runwasi/actions/runs/6191500343): wasmedge

**Comment by jprendes** [comment]
I'm hoping #323 can give us further insight into the issue

**Comment by jsturtevant** [comment]
Failed with wasmtime on this round: https://github.com/containerd/runwasi/actions/runs/6228704779/job/16906088261?pr=323

**Comment by dierbei** [comment]
@jsturtevant @jprendes I'll take a look at that.

Right now according to my thoughts it is:
1. look at the Containerd logs in Pending state
2. if all are in Pending state, are there any conflicts (e.g. as in unit tests)

**Comment by Mossaka** [comment]
Anyone tried reproducing it locally?

**Comment by dierbei** [comment]
> Anyone tried reproducing it locally?

Unfortunately, I tested make test-wasmer 13 times without any problems.

My OS is ubuntu 22.04.

**Comment by 0xE282B0** [comment]
Hi, I noticed that the first shim with sidecar that comes up after installation is stuck. When deleting it, it starts without problems.
Often it is Wasmer because it is the first one that starts but observed it with Lunatic and WasmEdge as well.
Like in this test report: https://github.com/KWasm/kwasm-node-installer/issues/43#issuecomment-1717317800

**Comment by 0xE282B0** [comment]
Not sure if it is the same problem but in my case I get this error message on the linux container whit `kubectl describe pod ...`:
```
Error: failed to create containerd task: 
  failed to start shim: start failed: io.containerd.wasmtime.v1: 
    Other("failed to setup namespaces: 
      Other: could not open network namespace /proc/0/ns/net: No such file or directory (os error 2)")
: exit status 1: unknown
```
Then the sidecar container is in a restart loop.

**Comment by dierbei** [comment]
I took a closer look at the logs and realized that it seems to be because `kubectl apply -f deploy.yaml` is executing too fast.

This is because the `--all-namespace` fetch does not fetch the pods in the k8s `kube-system` namespace.

```shell
sudo bin/k3s kubectl get pods --all-namespaces
NAMESPACE   NAME                        READY   STATUS    RESTARTS   AGE
default     wasi-demo-79d9475fd-p8r7k   0/2     Pending   0          107s
default     wasi-demo-79d9475fd-p7sqz   0/2     Pending   0          107s
default     wasi-demo-79d9475fd-c848k   0/2     Pending   0          107s
```

**Comment by jprendes** [comment]
> I took a closer look at the logs and realized that it seems to be because `kubectl apply -f deploy.yaml` is executing too fast.

I think you are right.
I've added some mitigation [here](https://github.com/containerd/runwasi/pull/323/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R164-R165), which runs k3s test for all runtimes:
* success: https://github.com/containerd/runwasi/actions/runs/6267907629/job/17021993601?pr=323
* success: https://github.com/containerd/runwasi/actions/runs/6267971481/job/17022065882?pr=323
* failed: https://github.com/containerd/runwasi/actions/runs/6268069976/job/17022318464?pr=323

That is compared to before, where I didn't manage to get a single clean run.

Now, if we check the failed run, it failed because the kube-system pods never came up after 1 minute, even before involving the shims, so it's not a problem with the shims.

**Comment by jsturtevant** [comment]
some additional logs investigations here: https://github.com/containerd/runwasi/pull/346#issuecomment-1745622322

**Comment by jsturtevant** [comment]
I've opened https://github.com/containerd/rust-extensions/pull/210 since our logs don't have timing info in them

**Comment by jsturtevant** [comment]
besides https://github.com/containerd/runwasi/pull/346#issuecomment-1745622322 seeing the following 

```
cp: cannot stat '/var/lib/rancher/k3s/agent/etc/containerd/config.toml': No such file or directory
```

**Comment by jprendes** [comment]
I think the best in that case, as wheel as when k3s's kube-system pods fall to star, is to uninstall k3s and install it again.

**Comment by jprendes** [comment]
This has been fixed by #353
